kratos/solving_strategies/builder_and_solvers/p_multigrid/constraint_assembler.hpp

```diff
@@ -0,0 +1,270 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+#pragma once
+
+// Project includes
+#include "solving_strategies/builder_and_solvers/p_multigrid/status_stream.hpp" // PMGStatusStream
+#include "containers/data_value_container.h" // DataValueContainer
+#include "containers/pointer_vector_set.h" // PointerVectorSet
+#include "includes/master_slave_constraint.h" // MasterSlaveConstraint
+#include "includes/indexed_object.h" // IndexedObject
+#include "includes/dof.h" // Dof
+#include "includes/variables.h" // IDENTIFIER
+#include "includes/smart_pointers.h" // KRATOS_CLASS_POINTER_DEFINITION
+
+
+namespace Kratos {
+
+
+/// @brief Enum class representing imposition methods for multifreedom constraints.
+enum class ConstraintImposition
+{
+    None                = 0,
+    MasterSlave         = 1,
+    Lagrange            = 2,
```

Comment:
Have your heard about double (not dual) LM?, it is better conditioned than LM, also bigger system. Just FYI